powerauth-java-server/src/main/java/com/wultra/security/powerauth/app/server/service/crypto/v4/CryptographyServiceHybrid.java

```diff
@@ -208,31 +250,120 @@ public void storeDevicePublicKey(ActivationRecordEntity activation, BasePublicKe
+                    final PrivateKey serverPrivateKey = privateKeyRegistry.getPrivateKey(KeyType.MLDSA_65).orElseThrow(() -> {
+                        logger.error("Missing server ML-DSA private key for activation ID: {}", activation.getActivationId());
+                        // Rollback is not required, database is not used for writing
+                        return localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+                    });
+                    yield PQC_DSA.sign(serverPrivateKey, data);
+                } catch (GenericCryptoException e) {
+                    logger.error("Could not generate signature", e);
+                    throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+                }
+            }
+            default -> {
+                logger.error("Unsupported key type in activation signature: {}", keyType);
+                throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+            }
+        };
     }
 
     @Override
-    public boolean verifySignatureForActivation(byte[] data, byte[] signature, ActivationRecordEntity activation) throws GenericServiceException {
-        // TODO
-        return false;
+    public boolean verifySignatureForActivation(KeyType keyType, byte[] data, byte[] signature, ActivationRecordEntity activation) throws GenericServiceException {
+        return switch (keyType) {
+            case ECDSA_P384 -> {
+                try {
+                    final PublicKeyRegistry publicKeyRegistry = publicKeysConverter.fromDBValue(activation.getDevicePublicKeys());
+                    final PublicKey devicePublicKey = publicKeyRegistry.getPublicKey(KeyType.ECDSA_P384).orElseThrow(() -> {
+                        logger.error("Missing device public key for ECDSA for activation ID: {}", activation.getActivationId());
+                        // Rollback is not required, database is not used for writing
+                        return localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+                    });
+                    yield SIGNATURE_UTILS.validateECDSASignature(EcCurve.P384, data, signature, devicePublicKey);
+                } catch (CryptoProviderException | GenericCryptoException | InvalidKeyException e) {
+                    logger.error("Invalid keypair", e);
+                    throw localizationProvider.buildExceptionForCode(ServiceError.GENERIC_CRYPTOGRAPHY_ERROR);
+                }
+            }
+            case MLDSA_65 -> {
+                try {
+                    final PublicKeyRegistry publicKeyRegistry = publicKeysConverter.fromDBValue(activation.getDevicePublicKeys());
+                    final PublicKey devicePublicKey = publicKeyRegistry.getPublicKey(KeyType.ECDSA_P384).orElseThrow(() -> {
```

Comment:
Wrong KeyType for the MLDSA signature
```suggestion
                    final PublicKey devicePublicKey = publicKeyRegistry.getPublicKey(KeyType.MLDSA_65).orElseThrow(() -> {
```

Reply:
Good catch!